Add integrity protection using Encrypt-Then-MAC to default encryption module

LukasReschke opened the pull request:
This adds an integrity protection using Encrypt-Then-MAC to the default encryption module. Authenticated encryption prevents that a storage administrator can tamper with the files such as performing bitflipping attacks on them.

To achieve this the following changes have been done by @schiesbn and me:
1. Switched to AES-CTR-256 as default cipher for new or modified files
2. All new files have the HMAC data in the metadata as well.
3. When decrypting files the signature gets verified. All files encrypted using CTR need this field to be there mandatory, the old CFB mode can have this field but it is not enforced.

This approach gives us the possibility to have complete backwards compatibility with the existing encrypted files (as in: you should be able to read all files as before as well) as well as an easy way to differentiate whether a file needs to have a mandatory HMAC.

To test this also try to modify the encrypted ciphertext in the encrypted document on new or modified files and see whether you can still open the document. With this change applied you should not, it will simply fail to decrypt the file.

There will be no automatic migration as we all know the pain of such. Thus this will only affect new or modified files.

Partially addresses https://github.com/owncloud/security-tracker/issues/191. Some documentation changes are still pending.

LukasReschke committed:
Use AES-256-CTR as default

CTR is recommended over CFB mode.

LukasReschke committed:
Use an actual 16 byte long IV

The previous IV was actually 12 byte extended to 16 byte using base64. As the encrypted file should be fine with containing binary data as well we can simply remove the encoding like that here.

LukasReschke committed:
Add note about the addPadding function

schiessle committed:
sign all encrypted blocks and check signature on decrypt

schiessle committed:
make it backward compatible to work with signed and un-signed files

schiessle committed:
always use default cipher for write operations, no matter how the file was encrypted before

schiessle committed:
meta data are at the end of the file

LukasReschke committed:
Use random_bytes instead OpenSSL

schiessle committed:
fixing unit tests

LukasReschke committed:
Use hash with appended "a" of the original password for the authentication

LukasReschke committed:
Clarify documentation

LukasReschke commented:
@owncloud/qa Please do some testing with encryption enabled. Thanks.
@schiesbn Our branch :tada:

karlitschek commented:
:+1:

rullzer commented:
So crypto stays a complex thing. And the more you learn about it the further you know to stay away from it ;).

Having said that I can follow the commits here. And the changes make sense to me.

LukasReschke commented:
Needs some more love to prevent shuffling the entries. Setting back to 2.

LukasReschke committed:
Keep track of file version

This way it is not possible anymore for an external storage admin to put up old versions of the file.

LukasReschke committed:
Use number of chunk for HMAC as well

Prevents switching single blocks within the encrypted file.

PVince81 commented:
- [ ] Blocked by: https://github.com/owncloud/core/issues/18255, cannot smash it

PVince81 commented:
- [x] TEST: regular encryption/decryption/overwrite
- [x] TEST: upgrade with old encrypted files, files can still be read
- [x] TEST: overwriting old encrypted files switches the encryption to the new mode

PVince81 commented:
Code looks good.

nickvergessen commented:
```
OCA\Files_sharing\Tests\EtagPropagation::testOwnerWritesToSingleFileShare
Failed asserting that the etag for "" of user test-share-user2 has changed
Failed asserting that 'eb98ef4f291e404506cbe04f7b577e83' is not equal to <string:eb98ef4f291e404506cbe04f7b577e83>.

/ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/apps/files_sharing/tests/propagationtestcase.php:63
/ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/apps/files_sharing/tests/etagpropagation.php:131
```

I saw a change in the error output with smashbox as well and this seems to confirm it... propagation is broken by this

LukasReschke committed:
Use database for keeping track of the version

PVince81 commented:
@schiesbn @LukasReschke please review each other's commits

LukasReschke commented:
> 05:51:39 1) OCA\Files_sharing\Tests\EtagPropagation::testOwnerWritesToSingleFileShare
05:51:39 Failed asserting that the etag for "" of user test-share-user2 has changed
05:51:39 Failed asserting that 'b2e5230ab95d4400c6683f5a6445745e' is not equal to <string:b2e5230ab95d4400c6683f5a6445745e>.

Unrelated…

LukasReschke commented:
:+1: for @schiesbn's changes.

nickvergessen commented:
Smashing: :hourglass_flowing_sand:

nickvergessen commented:
Smashbox fails :x: e.g. `lib/oc-tests/shareFile.py`

I used master and merged this PR into it.
Master passes, so this PR breaks something

Commit:
realPath should contain the path to the file we want to read, e.g. the version and not the original file

PVince81 commented:
```
% bin/smash -a lib/oc-tests/test_shareFile.py
SUMMARY:smash.:runid: 160209-174731
SUMMARY:smash.:URL: http://test:***@127.0.0.1/owncloud/remote.php/webdav/
SUMMARY:smash.:Running iteration 1
SUMMARY:smash.:running lib/oc-tests/test_shareFile.py in /home/vincent/work/smashdir/test_shareFile as test testset #0 {'test_sharePermissions': 31}
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
2016-02-09 17:48:23,335 - ERROR - shareeOne - File /home/vincent/work/smashdir/test_shareFile/shareeOne/TEST_FILE_USER_SHARE.dat should exist error_check(os.path.exists(sharedFile), "File %s should exist" %sharedFile) failed in shareeOne() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_shareFile.py" at line 158]
2016-02-09 17:48:23,336 - ERROR - shareeOne - File /home/vincent/work/smashdir/test_shareFile/shareeOne/TEST_FILE_USER_RESHARE.dat should exist error_check(os.path.exists(sharedFile), "File %s should exist" %sharedFile) failed in shareeOne() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_shareFile.py" at line 162]
2016-02-09 17:48:23,338 - ERROR - shareeOne - File /home/vincent/work/smashdir/test_shareFile/shareeOne/TEST_FILE_MODIFIED_USER_SHARE.dat should exist error_check(os.path.exists(sharedFile), "File %s should exist" %sharedFile) failed in shareeOne() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_shareFile.py" at line 166]
2016-02-09 17:48:23,349 - ERROR - shareeOne - /home/vincent/work/smashdir/test_shareFile/shareeOne/TEST_FILE_MODIFIED_USER_SHARE.dat does not exist
2016-02-09 17:48:24,475 - ERROR - sharer - md5 of modified file /home/vincent/work/smashdir/test_shareFile/sharer/TEST_FILE_MODIFIED_USER_SHARE.dat did not change: expected 45b565af28a2998fa8d5dae7eb6e660e "md5 of modified file %s did not change%s: expected %s" % (fn, comment, md5)) failed in expect_modified() ["/home/vincent/work/workspace/smashbox/python/smashbox/utilities/__init__.py" at line 781]
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
2016-02-09 17:48:26,653 - ERROR - shareeTwo - File /home/vincent/work/smashdir/test_shareFile/shareeTwo/TEST_FILE_USER_RESHARE.dat should exist error_check(os.path.exists(sharedFile), "File %s should exist" %sharedFile) failed in shareeTwo() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_shareFile.py" at line 229]
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
2016-02-09 17:48:28,740 - ERROR - shareeTwo - File /home/vincent/work/smashdir/test_shareFile/shareeTwo/TEST_FILE_USER_RESHARE.dat does not exist but should error_check(os.path.exists(fn), "File %s does not exist but should" % fn) failed in expect_exists() ["/home/vincent/work/workspace/smashbox/python/smashbox/utilities/__init__.py" at line 795]
2016-02-09 17:48:31,241 - ERROR - sharer - 1 error(s) reported
2016-02-09 17:48:31,241 - ERROR - shareeTwo - 2 error(s) reported
2016-02-09 17:48:31,241 - ERROR - shareeOne - 4 error(s) reported
SUMMARY:smash.:Elapsed time: 59s (0:00:59.322018)
CRITICAL:smash.:Aborting run -- non-zero exit code (2)
```

PVince81 commented:
@schiesbn doesn't seem to fix it ^

PVince81 commented:
Steps to reproduce:

1. user1 shares file "test.txt" with user2
1. as user2, do a Webdav PUT to overwrite the file

Stack trace:
```
{"reqId":"YQ88SeNubNctmbfEMcbt","remoteAddr":"127.0.0.1","app":"PHP","message":"Call to a member function getData() on array at \/srv\/www\/htdocs\/owncloud\/lib\/private\/files\/storage\/wrapper\/encryption.php#133","level":3,"time":"2016-02-09T16:56:49+00:00","method":"PUT","url":"\/owncloud\/remote.php\/webdav\/test.txt"}
```

Looks like the exact same issue I fixed yesterday on master. Let me recheck the code.

Commit:
Check if partial cache entry or not in encryption wrapper

PVince81 commented:
Okay, so there was a new code path where the array got re-read.
I fixed the issue: https://github.com/owncloud/core/commit/397845f10c6e226ccddcbc0545d6eadca3719a44

However, now we got the next smashbox fail:
```
± % bin/smash -a lib/oc-tests/test_shareFile.py                                                                                                                                                                                      !11037
SUMMARY:smash.:runid: 160209-180851
SUMMARY:smash.:URL: http://test:***@127.0.0.1/owncloud/remote.php/webdav/
SUMMARY:smash.:Running iteration 1
SUMMARY:smash.:running lib/oc-tests/test_shareFile.py in /home/vincent/work/smashdir/test_shareFile as test testset #0 {'test_sharePermissions': 31}
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
2016-02-09 18:09:36,366 - ERROR - shareeOne - File /home/vincent/work/smashdir/test_shareFile/shareeOne/TEST_FILE_USER_SHARE.dat should exist error_check(os.path.exists(sharedFile), "File %s should exist" %sharedFile) failed in shareeOne() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_shareFile.py" at line 158]
2016-02-09 18:09:36,367 - ERROR - shareeOne - File /home/vincent/work/smashdir/test_shareFile/shareeOne/TEST_FILE_USER_RESHARE.dat should exist error_check(os.path.exists(sharedFile), "File %s should exist" %sharedFile) failed in shareeOne() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_shareFile.py" at line 162]
2016-02-09 18:09:36,368 - ERROR - shareeOne - File /home/vincent/work/smashdir/test_shareFile/shareeOne/TEST_FILE_MODIFIED_USER_SHARE.dat should exist error_check(os.path.exists(sharedFile), "File %s should exist" %sharedFile) failed in shareeOne() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_shareFile.py" at line 166]
2016-02-09 18:09:36,378 - ERROR - shareeOne - /home/vincent/work/smashdir/test_shareFile/shareeOne/TEST_FILE_MODIFIED_USER_SHARE.dat does not exist
2016-02-09 18:09:37,532 - ERROR - sharer - md5 of modified file /home/vincent/work/smashdir/test_shareFile/sharer/TEST_FILE_MODIFIED_USER_SHARE.dat did not change: expected 45b565af28a2998fa8d5dae7eb6e660e "md5 of modified file %s did not change%s: expected %s" % (fn, comment, md5)) failed in expect_modified() ["/home/vincent/work/workspace/smashbox/python/smashbox/utilities/__init__.py" at line 781]
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
2016-02-09 18:09:39,689 - ERROR - shareeTwo - File /home/vincent/work/smashdir/test_shareFile/shareeTwo/TEST_FILE_USER_RESHARE.dat should exist error_check(os.path.exists(sharedFile), "File %s should exist" %sharedFile) failed in shareeTwo() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_shareFile.py" at line 229]
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1
2016-02-09 18:09:41,581 - ERROR - shareeTwo - File /home/vincent/work/smashdir/test_shareFile/shareeTwo/TEST_FILE_USER_RESHARE.dat does not exist but should error_check(os.path.exists(fn), "File %s does not exist but should" % fn) failed in expect_exists() ["/home/vincent/work/workspace/smashbox/python/smashbox/utilities/__init__.py" at line 795]
2016-02-09 18:09:44,207 - ERROR - sharer - 1 error(s) reported
2016-02-09 18:09:44,213 - ERROR - shareeOne - 4 error(s) reported
2016-02-09 18:09:44,215 - ERROR - shareeTwo - 2 error(s) reported
SUMMARY:smash.:Elapsed time: 52s (0:00:52.286788)
CRITICAL:smash.:Aborting run -- non-zero exit code (2)
```

And matching logs with BadSignature!! :scream: 

```
{"reqId":"BuIVfe5H\/utatG9IPOnz","remoteAddr":"127.0.0.1","app":"ocs_api","message":"Successful addUser call with userid: test3","level":1,"time":"2016-02-09T17:09:25+00:00","method":"POST","url":"\/owncloud\/ocs\/v1.php\/cloud\/users"}
{"reqId":"ur4QeAqLCwV\/brUJiPm9","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"Bad Signature\",\"Exception\":\"OC\\\\HintException\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php(457): OCA\\\\Encryption\\\\Crypto\\\\Crypt->checkSignature('\\\/+tEnjqiVhv6ki3...', '\\\\xCD}B\\\\xF1\\\\xF3\\\\xE5\\\\xF5J\\\\xED\\\\xF7\\\\xBF(\\\\xFE\\\\x1C~...', '1e0cef1488089f0...')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/encryption.php(354): OCA\\\\Encryption\\\\Crypto\\\\Crypt->symmetricDecryptFileContent('\\\/+tEnjqiVhv6ki3...', '\\\\xCD}B\\\\xF1\\\\xF3\\\\xE5\\\\xF5J\\\\xED\\\\xF7\\\\xBF(\\\\xFE\\\\x1C~...', 'AES-256-CTR', 1, '1end')\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(458): OCA\\\\Encryption\\\\Crypto\\\\Encryption->decrypt('\\\/+tEnjqiVhv6ki3...', '1end')\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(289): OC\\\\Files\\\\Stream\\\\Encryption->readCache()\\n#4 [internal function]: OC\\\\Files\\\\Stream\\\\Encryption->stream_read(8192)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/Wrapper.php(83): fread(Resource id #427, 8192)\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/CallbackWrapper.php(91): Icewind\\\\Streams\\\\Wrapper->stream_read(8192)\\n#7 [internal function]: Icewind\\\\Streams\\\\CallbackWrapper->stream_read(8192)\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/http\\\/lib\\\/Sapi.php(78): stream_copy_to_stream(Resource id #431, Resource id #441, '10000')\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(470): Sabre\\\\HTTP\\\\Sapi::sendResponse(Object(Sabre\\\\HTTP\\\\Response))\\n#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(54): Sabre\\\\DAV\\\\Server->exec()\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(137): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#13 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php\",\"Line\":477}","level":4,"time":"2016-02-09T17:09:36+00:00","method":"GET","url":"\/owncloud\/remote.php\/webdav\/TEST_FILE_MODIFIED_USER_SHARE.dat"}
{"reqId":"zgKlfJa4QO+xXb\/4jHK2","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"Bad Signature\",\"Exception\":\"OC\\\\HintException\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php(457): OCA\\\\Encryption\\\\Crypto\\\\Crypt->checkSignature('weqf0k5vXzfnfmW...', '\\\\x91\\\\xC7\\\\x83\\\\xD2U\\\\xFB\\\\x8E\\\\xFBR\\\\xF6\\\\xF7a\\\\x89E\\\\x01...', '266c9a3df9391bf...')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/encryption.php(354): OCA\\\\Encryption\\\\Crypto\\\\Crypt->symmetricDecryptFileContent('weqf0k5vXzfnfmW...', '\\\\x91\\\\xC7\\\\x83\\\\xD2U\\\\xFB\\\\x8E\\\\xFBR\\\\xF6\\\\xF7a\\\\x89E\\\\x01...', 'AES-256-CTR', 1, '1end')\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(458): OCA\\\\Encryption\\\\Crypto\\\\Encryption->decrypt('weqf0k5vXzfnfmW...', '1end')\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(289): OC\\\\Files\\\\Stream\\\\Encryption->readCache()\\n#4 [internal function]: OC\\\\Files\\\\Stream\\\\Encryption->stream_read(8192)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/Wrapper.php(83): fread(Resource id #427, 8192)\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/CallbackWrapper.php(91): Icewind\\\\Streams\\\\Wrapper->stream_read(8192)\\n#7 [internal function]: Icewind\\\\Streams\\\\CallbackWrapper->stream_read(8192)\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/http\\\/lib\\\/Sapi.php(78): stream_copy_to_stream(Resource id #431, Resource id #441, '10000')\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(470): Sabre\\\\HTTP\\\\Sapi::sendResponse(Object(Sabre\\\\HTTP\\\\Response))\\n#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(54): Sabre\\\\DAV\\\\Server->exec()\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(137): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#13 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php\",\"Line\":477}","level":4,"time":"2016-02-09T17:09:36+00:00","method":"GET","url":"\/owncloud\/remote.php\/webdav\/TEST_FILE_USER_RESHARE.dat"}
{"reqId":"dcEOMCxdR7Tmxz8FHnke","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"Bad Signature\",\"Exception\":\"OC\\\\HintException\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php(457): OCA\\\\Encryption\\\\Crypto\\\\Crypt->checkSignature('Kc34G2\\\/nsLv0mJl...', 'zW\\\\x8F\\\\xBDG\\\\x8C@\\\\xBA\\\\xC7\\\\xA3&\\\\xE9\\\\xE8Ex...', 'd9024a0da4fe4ec...')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/encryption.php(354): OCA\\\\Encryption\\\\Crypto\\\\Crypt->symmetricDecryptFileContent('Kc34G2\\\/nsLv0mJl...', 'zW\\\\x8F\\\\xBDG\\\\x8C@\\\\xBA\\\\xC7\\\\xA3&\\\\xE9\\\\xE8Ex...', 'AES-256-CTR', 1, '1end')\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(458): OCA\\\\Encryption\\\\Crypto\\\\Encryption->decrypt('Kc34G2\\\/nsLv0mJl...', '1end')\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(289): OC\\\\Files\\\\Stream\\\\Encryption->readCache()\\n#4 [internal function]: OC\\\\Files\\\\Stream\\\\Encryption->stream_read(8192)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/Wrapper.php(83): fread(Resource id #427, 8192)\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/CallbackWrapper.php(91): Icewind\\\\Streams\\\\Wrapper->stream_read(8192)\\n#7 [internal function]: Icewind\\\\Streams\\\\CallbackWrapper->stream_read(8192)\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/http\\\/lib\\\/Sapi.php(78): stream_copy_to_stream(Resource id #431, Resource id #441, '10000')\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(470): Sabre\\\\HTTP\\\\Sapi::sendResponse(Object(Sabre\\\\HTTP\\\\Response))\\n#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(54): Sabre\\\\DAV\\\\Server->exec()\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(137): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#13 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php\",\"Line\":477}","level":4,"time":"2016-02-09T17:09:36+00:00","method":"GET","url":"\/owncloud\/remote.php\/webdav\/TEST_FILE_USER_SHARE.dat"}
{"reqId":"nnbETGuzbAm1L8a\/IJ6r","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"Bad Signature\",\"Exception\":\"OC\\\\HintException\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php(457): OCA\\\\Encryption\\\\Crypto\\\\Crypt->checkSignature('weqf0k5vXzfnfmW...', '\\\\x91\\\\xC7\\\\x83\\\\xD2U\\\\xFB\\\\x8E\\\\xFBR\\\\xF6\\\\xF7a\\\\x89E\\\\x01...', '266c9a3df9391bf...')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/encryption.php(354): OCA\\\\Encryption\\\\Crypto\\\\Crypt->symmetricDecryptFileContent('weqf0k5vXzfnfmW...', '\\\\x91\\\\xC7\\\\x83\\\\xD2U\\\\xFB\\\\x8E\\\\xFBR\\\\xF6\\\\xF7a\\\\x89E\\\\x01...', 'AES-256-CTR', 1, '1end')\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(458): OCA\\\\Encryption\\\\Crypto\\\\Encryption->decrypt('weqf0k5vXzfnfmW...', '1end')\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(289): OC\\\\Files\\\\Stream\\\\Encryption->readCache()\\n#4 [internal function]: OC\\\\Files\\\\Stream\\\\Encryption->stream_read(8192)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/Wrapper.php(83): fread(Resource id #427, 8192)\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/CallbackWrapper.php(91): Icewind\\\\Streams\\\\Wrapper->stream_read(8192)\\n#7 [internal function]: Icewind\\\\Streams\\\\CallbackWrapper->stream_read(8192)\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/http\\\/lib\\\/Sapi.php(78): stream_copy_to_stream(Resource id #431, Resource id #441, '10000')\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(470): Sabre\\\\HTTP\\\\Sapi::sendResponse(Object(Sabre\\\\HTTP\\\\Response))\\n#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(54): Sabre\\\\DAV\\\\Server->exec()\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(137): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#13 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/crypt.php\",\"Line\":477}","level":4,"time":"2016-02-09T17:09:39+00:00","method":"GET","url":"\/owncloud\/remote.php\/webdav\/TEST_FILE_USER_RESHARE.dat"}
```

@LukasReschke @schiesbn

PVince81 commented:
(sorry for the master merge... was in a rush and wanted to test with master code applied, forgot to revert)

PVince81 commented:
Steps for the bad signature:

1. Create a local file that has 8kb or more (I used `dd if=/dev/urandom of=8192.dat count=1 bs=8192`)
1. Upload the file as "user1"
1. Share the file with "user2"
1. Using Webdav as "user2", PUT the same file again and overwrite the original one (from the same original file)

```
{"reqId":"4TxIiQrxM1saPrzdOkEd","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 500 Bad Signature\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/connector\\\/sabre\\\/file.php(193): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->convertToSabreException(Object(OC\\\\HintException))\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1070): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->put(Resource id #302)\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(513): Sabre\\\\DAV\\\\Server->updateFile('8192.dat', Resource id #302, NULL)\\n#3 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPut(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(459): Sabre\\\\Event\\\\EventEmitter->emit('method:PUT', Array)\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(54): Sabre\\\\DAV\\\\Server->exec()\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(137): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#9 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/connector\\\/sabre\\\/file.php\",\"Line\":535}","level":4,"time":"2016-02-09T17:17:20+00:00","method":"PUT","url":"\/owncloud\/remote.php\/webdav\/8192.dat"}
```

Actually, it seems that overwriting ANY file (even small) with a big one fails.

PVince81 commented:
Seems to work fine with folder shares. Only file share + overwrite + big file seem to be affected.

PVince81 commented:
Looks like downloading the file as user2 also fails (at least after trying to overwrite it).
The reason the upload fails is because it internally does a stream copy (copyFromStorage?!) so it needs to read the part file, decrypt it, then reencrypt it to final storage.

PVince81 commented:
I tried disabling signature check, the download works. The md5 of the file is correct.
So it's not an issue with file integrity.

Commit:
don't decrease ->version for part files but only a local variable, otherwise it can happen that we decrease it twice and end up with the wrong value

schiesbn commented:
@PVince81 @LukasReschke my last commit should fix it.

Background: when writing the chunk file we decreased `$this->version` in `encrypt()` by writing multiple chunks we decreased it for every chunk so the number get lower and lower. Which if course doesn't match the expected value if we read the part file later again. By using a local variable for the version we make sure that we use the same version number for each chunk.

schiesbn commented on apps/encryption/lib/keymanager.php:
```diff
@@ -413,6 +415,41 @@ public function getFileKey($path, $uid) {
+
+		if($fileInfo !== false) {
+			$fileId = $fileInfo->getId();
+			$qb = \OC::$server->getDatabaseConnection()->getQueryBuilder();
```
do we really want to manipulate the file cache directly? Wouldn't it be better to use ```$cache->put($path, ['encrypted' => $version]);  ?

cc @icewind1991

PVince81 replied:
right, this has to go through the Cache API because some caches might store the information elsewhere.

You should be able to retrieve it from the view: `$fileInfo->getStorage()->getCache()`

PVince81 replied:
@schiesbn can you make the change ?

schiesbn commented on apps/encryption/lib/keymanager.php:
```diff
@@ -413,6 +415,41 @@ public function getFileKey($path, $uid) {
+	 * @return int
+	 */
+	public function getVersion($path) {
+		$view = new \OC\Files\View();
```
would be nice to inject the $view, this way we can write tests for the method

LukasReschke replied:
Good point. I'll do that.

schiesbn commented on apps/encryption/lib/keymanager.php:
```diff
@@ -413,6 +415,41 @@ public function getFileKey($path, $uid) {
+	 * @param string $version
+	 */
+	public function setVersion($path, $version) {
+		$view = new \OC\Files\View();
```
same as above

PVince81 commented:
I'll resmash this.

PVince81 commented:
ah, test_shareFile passed! Now let's do the full smash.

PVince81 commented:
Fullsmash passed this time, yay !

Before we can merge this, we need to fix the cache access here: https://github.com/owncloud/core/pull/21557#discussion_r52358464

LukasReschke committed:
Use cache and add tests

Commit:
Use cache update instead of put for encryption version

Saves a call to fetch the file id which didn't even work for a reason.

This fix properly sets the version in the database.

LukasReschke committed:
Add tests for setVersion

PVince81 commented on apps/encryption/lib/crypto/encryption.php:
```diff
@@ -204,17 +220,25 @@ public function begin($path, $user, $mode, array $header, array $accessList) {
+			} else {
+				$version = $this->version + 1;
+			}
+			$this->keyManager->setVersion($this->path, $this->version+1, new View());
```
not knowing the algo very well, I assume this `$this->version + 1` is correct here and is independent from the part file logic, correct ?

schiesbn replied:
hm, I would say that this is wrong, we should use here just `$version`. But all tests passed and all manual tests I did worked as well... @LukasReschke what do you think?

LukasReschke replied:
We had both set to `$this->version + 1` before https://github.com/owncloud/core/commit/377d7fb8a817f337552e444c5aea9eb82bcfada4. Let me think what the impact would be here.

PVince81 commented:
Code looks good, smashbox tests passed locally with the latest changes :+1:

LukasReschke commented:
:+1: for @schiesbn's changes from my side.
